File: runner/runner.go
package runner

import (
	"errors"
	"fmt"
	"maps"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"sync"

	"github.com/infraflakes/sro/ast"
	"github.com/infraflakes/sro/config"
)

type Runner struct {
	cfg *config.Config
}

func New(cfg *config.Config) *Runner {
	return &Runner{cfg: cfg}
}

type execContext struct {
	cfg      *config.Config
	project  *config.Project
	vars     map[string]string
	envStack []map[string]string
	workDir  string
}

func newExecContext(cfg *config.Config, proj *config.Project) *execContext {
	vars := make(map[string]string, len(cfg.Vars))
	maps.Copy(vars, cfg.Vars)

	baseDir := filepath.Join(cfg.Sanctuary, proj.Dir)

	return &execContext{
		cfg:      cfg,
		project:  proj,
		vars:     vars,
		envStack: []map[string]string{},
		workDir:  baseDir,
	}
}

func (ctx *execContext) resolveArgs(args []ast.Expr) ([]string, error) {
	result := make([]string, 0, len(args))
	for _, arg := range args {
		switch a := arg.(type) {
		case *ast.StringLit:
			result = append(result, a.Value)
		case *ast.VarRef:
			val, ok := ctx.vars[a.Name]
			if !ok {
				line, col := a.Pos()
				return nil, fmt.Errorf("%d:%d: undefined variable $%s", line, col, a.Name)
			}
			result = append(result, val)
		default:
			return nil, fmt.Errorf("unexpected expression type: %T", arg)
		}
	}
	return result, nil
}

func (ctx *execContext) buildEnv() []string {
	env := map[string]string{}
	for _, e := range os.Environ() {
		parts := strings.SplitN(e, "=", 2)
		if len(parts) == 2 {
			env[parts[0]] = parts[1]
		}
	}

	for _, layer := range ctx.envStack {
		maps.Copy(env, layer)
	}

	result := make([]string, 0, len(env))
	for k, v := range env {
		result = append(result, k+"="+v)
	}
	return result
}

func (ctx *execContext) execFnBody(body []ast.FnStmt) error {
	for _, stmt := range body {
		switch s := stmt.(type) {
		case *ast.LogStmt:
			if err := ctx.execLog(s); err != nil {
				return err
			}
		case *ast.ExecStmt:
			if err := ctx.execExec(s); err != nil {
				return err
			}
		case *ast.CdStmt:
			if err := ctx.execCd(s); err != nil {
				return err
			}
		case *ast.VarDecl:
			if err := ctx.execVarDecl(s); err != nil {
				return err
			}
		case *ast.EnvBlock:
			if err := ctx.execEnvBlock(s); err != nil {
				return err
			}
		default:
			return fmt.Errorf("unknown fn statement: %T", stmt)
		}
	}
	return nil
}

func (ctx *execContext) execLog(s *ast.LogStmt) error {
	args, err := ctx.resolveArgs(s.Args)
	if err != nil {
		return err
	}
	fmt.Println(strings.Join(args, " "))
	return nil
}

func (ctx *execContext) execExec(s *ast.ExecStmt) error {
	args, err := ctx.resolveArgs(s.Args)
	if err != nil {
		return err
	}
	cmdStr := strings.Join(args, " ")
	fmt.Printf("  exec  %s\n", cmdStr)

	cmd := exec.Command("sh", "-c", cmdStr)
	cmd.Dir = ctx.workDir
	cmd.Env = ctx.buildEnv()
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr

	if err := cmd.Run(); err != nil {
		var exitErr *exec.ExitError
		if errors.As(err, &exitErr) {
			return fmt.Errorf("exec failed with exit code %d: %s", exitErr.ExitCode(), cmdStr)
		}
		return fmt.Errorf("exec failed: %s: %w", cmdStr, err)
	}
	return nil
}

func (ctx *execContext) execCd(s *ast.CdStmt) error {
	baseDir := filepath.Join(ctx.cfg.Sanctuary, ctx.project.Dir)
	if s.Arg == "." {
		ctx.workDir = baseDir
	} else {
		ctx.workDir = filepath.Join(baseDir, s.Arg)
	}
	if _, err := os.Stat(ctx.workDir); err != nil {
		return fmt.Errorf("cd %q: %w", s.Arg, err)
	}
	return nil
}

func (ctx *execContext) execVarDecl(s *ast.VarDecl) error {
	switch v := s.Value.(type) {
	case *ast.StringLit:
		ctx.vars[s.Name] = v.Value
	case *ast.VarRef:
		val, ok := ctx.vars[v.Name]
		if !ok {
			return fmt.Errorf("undefined variable: $%s", v.Name)
		}
		ctx.vars[s.Name] = val
	default:
		return fmt.Errorf("unexpected value type: %T", v)
	}
	return nil
}

func (ctx *execContext) execEnvBlock(s *ast.EnvBlock) error {
	layer := make(map[string]string, len(s.Pairs))
	for _, p := range s.Pairs {
		switch v := p.Value.(type) {
		case *ast.StringLit:
			layer[p.Key] = v.Value
		case *ast.VarRef:
			val, ok := ctx.vars[v.Name]
			if !ok {
				return fmt.Errorf("undefined variable: $%s", v.Name)
			}
			layer[p.Key] = val
		}
	}
	ctx.envStack = append(ctx.envStack, layer)

	// Snapshot vars before body so local vars don't leak
	savedVars := make(map[string]string, len(ctx.vars))
	maps.Copy(savedVars, ctx.vars)

	err := ctx.execFnBody(s.Body)

	ctx.vars = savedVars
	ctx.envStack = ctx.envStack[:len(ctx.envStack)-1]
	return err
}

func (r *Runner) RunSeq(name string) error {
	seq, ok := r.cfg.Seqs[name]
	if !ok {
		return fmt.Errorf("unknown seq: %s", name)
	}
	fmt.Printf("seq %s\n", seq.Name)
	return r.executeSeq(seq)
}

func (r *Runner) executeSeq(seq *ast.SeqDecl) error {
	for _, stmt := range seq.Stmts {
		switch s := stmt.(type) {
		case *ast.FnCall:
			if err := r.executeFnCall(s); err != nil {
				return err
			}
		case *ast.SeqRef:
			refSeq, ok := r.cfg.Seqs[s.SeqName]
			if !ok {
				return fmt.Errorf("unknown seq: %s", s.SeqName)
			}
			if err := r.executeSeq(refSeq); err != nil {
				return err
			}
		default:
			return fmt.Errorf("unknown seq statement: %T", stmt)
		}
	}
	return nil
}

func (r *Runner) executeFnCall(call *ast.FnCall) error {
	fn, ok := r.cfg.Functions[call.FnName]
	if !ok {
		return fmt.Errorf("unknown function: %s", call.FnName)
	}
	proj, ok := r.cfg.Projects[call.ProjectName]
	if !ok {
		return fmt.Errorf("unknown project: %s", call.ProjectName)
	}

	fmt.Printf("  %s(pr.%s)\n", call.FnName, call.ProjectName)
	ctx := newExecContext(r.cfg, proj)
	return ctx.execFnBody(fn.Body)
}

func (r *Runner) RunPar(name string) error {
	par, ok := r.cfg.Pars[name]
	if !ok {
		return fmt.Errorf("unknown par: %s", name)
	}

	fmt.Printf("par %s\n", par.Name)

	var wg sync.WaitGroup
	var mu sync.Mutex
	var errs []error

	for _, stmt := range par.Stmts {
		switch s := stmt.(type) {
		case *ast.FnCall:
			wg.Add(1)
			go func(call *ast.FnCall) {
				defer wg.Done()
				if err := r.executeFnCall(call); err != nil {
					mu.Lock()
					errs = append(errs, fmt.Errorf("%s(pr.%s): %w", call.FnName, call.ProjectName, err))
					mu.Unlock()
				}
			}(s)
		case *ast.SeqRef:
			wg.Add(1)
			go func(ref *ast.SeqRef) {
				defer wg.Done()
				refSeq, ok := r.cfg.Seqs[ref.SeqName]
				if !ok {
					mu.Lock()
					errs = append(errs, fmt.Errorf("unknown seq: %s", ref.SeqName))
					mu.Unlock()
					return
				}
				if err := r.executeSeq(refSeq); err != nil {
					mu.Lock()
					errs = append(errs, fmt.Errorf("seq.%s: %w", ref.SeqName, err))
					mu.Unlock()
				}
			}(s)
		default:
			mu.Lock()
			errs = append(errs, fmt.Errorf("unknown par statement: %T", stmt))
			mu.Unlock()
		}
	}

	wg.Wait()

	if len(errs) > 0 {
		msgs := make([]string, len(errs))
		for i, e := range errs {
			msgs[i] = e.Error()
		}
		return fmt.Errorf("par %s: %d error(s):\n%s", name, len(errs), strings.Join(msgs, "\n"))
	}
	return nil
}

File: src/runner.rs
use crate::ast::{FnStmt, SeqStmt, ParStmt, Expr};
use crate::config::{Config, Project, ConfigError};
use std::collections::HashMap;
use std::io::{self, Write};
use std::path::PathBuf;
use std::process::Command;

pub struct Runner {
    cfg: Config,
    writer: Box<dyn Write>,
    suppress_headers: bool,
}

impl Runner {
    pub fn new(cfg: Config) -> Self {
        Runner {
            cfg,
            writer: Box::new(io::stdout()),
            suppress_headers: false,
        }
    }

    #[allow(dead_code)]
    pub fn with_writer(mut self, writer: Box<dyn Write>) -> Self {
        self.writer = writer;
        self
    }

    #[allow(dead_code)]
    pub fn suppress_headers(mut self, suppress: bool) -> Self {
        self.suppress_headers = suppress;
        self
    }

    pub fn execute_fn_call(&mut self, fn_name: &str, project_name: &str) -> Result<(), ConfigError> {
        let fn_decl = self.cfg.functions.get(fn_name)
            .ok_or_else(|| ConfigError::Validation(format!("unknown function: {}", fn_name)))?
            .clone();
        
        let project = self.cfg.projects.get(project_name)
            .ok_or_else(|| ConfigError::Validation(format!("unknown project: {}", project_name)))?
            .clone();

        if !self.suppress_headers {
            writeln!(self.writer, "\x1b[38;2;91;156;246m{}\x1b[0m({})", fn_name, project_name)
                .map_err(|e| ConfigError::Validation(format!("write error: {}", e)))?;
        }

        let mut ctx = ExecContext::new(&self.cfg, &project, &mut *self.writer);
        if let crate::ast::Stmt::FnDecl { body, .. } = &fn_decl {
            ctx.exec_fn_body(body)?;
        }
        Ok(())
    }

    pub fn run_seq(&mut self, seq_name: &str) -> Result<(), ConfigError> {
        let seq_decl = self.cfg.seqs.get(seq_name)
            .ok_or_else(|| ConfigError::Validation(format!("unknown seq: {}", seq_name)))?
            .clone();

        if !self.suppress_headers {
            writeln!(self.writer, "seq {}", seq_name)
                .map_err(|e| ConfigError::Validation(format!("write error: {}", e)))?;
        }

        self.execute_seq(&seq_decl)
    }

    fn execute_seq(&mut self, seq_decl: &crate::ast::Stmt) -> Result<(), ConfigError> {
        if let crate::ast::Stmt::SeqDecl { stmts, .. } = seq_decl {
            for stmt in stmts {
                match stmt {
                    SeqStmt::FnCall { fn_name, project_name, .. } => {
                        self.execute_fn_call(fn_name, project_name)?;
                    }
                    SeqStmt::SeqRef { seq_name, .. } => {
                        let ref_seq = self.cfg.seqs.get(seq_name)
                            .ok_or_else(|| ConfigError::Validation(format!("unknown seq: {}", seq_name)))?
                            .clone();
                        self.execute_seq(&ref_seq)?;
                    }
                }
            }
        }
        Ok(())
    }

    pub fn run_par(&mut self, par_name: &str) -> Result<(), ConfigError> {
        let par_decl = self.cfg.pars.get(par_name)
            .ok_or_else(|| ConfigError::Validation(format!("unknown par: {}", par_name)))?
            .clone();

        if !self.suppress_headers {
            writeln!(self.writer, "par {}", par_name)
                .map_err(|e| ConfigError::Validation(format!("write error: {}", e)))?;
        }

        self.execute_par(&par_decl)
    }

    fn execute_par(&mut self, par_decl: &crate::ast::Stmt) -> Result<(), ConfigError> {
        // For now, execute sequentially - will need async for true parallel execution
        if let crate::ast::Stmt::ParDecl { stmts, .. } = par_decl {
            for stmt in stmts {
                match stmt {
                    ParStmt::FnCall { fn_name, project_name, .. } => {
                        self.execute_fn_call(fn_name, project_name)?;
                    }
                    ParStmt::SeqRef { seq_name, .. } => {
                        let ref_seq = self.cfg.seqs.get(seq_name)
                            .ok_or_else(|| ConfigError::Validation(format!("unknown seq: {}", seq_name)))?
                            .clone();
                        self.execute_seq(&ref_seq)?;
                    }
                }
            }
        }
        Ok(())
    }
}

struct ExecContext<'a> {
    cfg: &'a Config,
    project: &'a Project,
    vars: HashMap<String, String>,
    env_stack: Vec<HashMap<String, String>>,
    work_dir: PathBuf,
    writer: &'a mut dyn Write,
}

impl<'a> ExecContext<'a> {
    fn new(cfg: &'a Config, project: &'a Project, writer: &'a mut dyn Write) -> Self {
        let mut vars = HashMap::new();
        for (k, v) in &cfg.vars {
            vars.insert(k.clone(), v.clone());
        }

        let base_dir = PathBuf::from(&cfg.sanctuary).join(&project.dir);

        ExecContext {
            cfg,
            project,
            vars,
            env_stack: Vec::new(),
            work_dir: base_dir,
            writer,
        }
    }

    fn exec_fn_body(&mut self, body: &[FnStmt]) -> Result<(), ConfigError> {
        for stmt in body {
            match stmt {
                FnStmt::Log { value, .. } => self.exec_log(value)?,
                FnStmt::Exec { value, .. } => self.exec_exec(value)?,
                FnStmt::Cd { arg, .. } => self.exec_cd(arg)?,
                FnStmt::VarDecl { name, value, var_type, .. } => self.exec_var_decl(name, value, var_type)?,
                FnStmt::EnvBlock { pairs, body: block_body, .. } => self.exec_env_block(pairs, block_body)?,
            }
        }
        Ok(())
    }

    fn exec_log(&mut self, value: &Expr) -> Result<(), ConfigError> {
        let msg = self.resolve_expr(value)?;
        let indent = "  ".repeat(self.env_stack.len());
        writeln!(self.writer, "{}\x1b[38;2;255;203;107mlog  {}\x1b[0m", indent, msg)
            .map_err(|e| ConfigError::Validation(format!("write error: {}", e)))?;
        Ok(())
    }

    fn exec_exec(&mut self, value: &Expr) -> Result<(), ConfigError> {
        let cmd_str = self.resolve_expr(value)?;
        let indent = "  ".repeat(self.env_stack.len());
        writeln!(self.writer, "{}\x1b[38;2;91;156;246mexec {}\x1b[0m", indent, cmd_str)
            .map_err(|e| ConfigError::Validation(format!("write error: {}", e)))?;

        let output = Command::new(&self.cfg.shell)
            .arg("-c")
            .arg(&cmd_str)
            .current_dir(&self.work_dir)
            .envs(self.build_env())
            .output()
            .map_err(|e| ConfigError::Validation(format!("exec failed: {}: {}", cmd_str, e)))?;

        if !output.status.success() {
            return Err(ConfigError::Validation(format!(
                "exec failed with exit code: {}",
                output.status.code().unwrap_or(-1)
            )));
        }

        let stdout = String::from_utf8_lossy(&output.stdout);
        let stderr = String::from_utf8_lossy(&output.stderr);
        
        if !stdout.is_empty() {
            let stdout_indent = "  ".repeat(self.env_stack.len() + 1);
            for line in stdout.lines() {
                writeln!(self.writer, "{}{}", stdout_indent, line)
                    .map_err(|e| ConfigError::Validation(format!("write error: {}", e)))?;
            }
        }
        
        if !stderr.is_empty() {
            let stderr_indent = "  ".repeat(self.env_stack.len() + 1);
            for line in stderr.lines() {
                writeln!(self.writer, "{}{}", stderr_indent, line)
                    .map_err(|e| ConfigError::Validation(format!("write error: {}", e)))?;
            }
        }

        Ok(())
    }

    fn exec_cd(&mut self, arg: &str) -> Result<(), ConfigError> {
        let base_dir = PathBuf::from(&self.cfg.sanctuary).join(&self.project.dir);
        self.work_dir = if arg == "." {
            base_dir
        } else {
            base_dir.join(arg)
        };

        if !self.work_dir.exists() {
            return Err(ConfigError::Validation(format!("cd {}: directory does not exist", arg)));
        }

        let indent = "  ".repeat(self.env_stack.len());
        writeln!(self.writer, "{}\x1b[38;2;255;203;107mcd   {}\x1b[0m", indent, arg)
            .map_err(|e| ConfigError::Validation(format!("write error: {}", e)))?;
        Ok(())
    }

    fn exec_var_decl(&mut self, name: &str, value: &Expr, var_type: &crate::ast::VarType) -> Result<(), ConfigError> {
        let val = self.resolve_expr(value)?;
        
        if var_type == &crate::ast::VarType::Shell {
            let output = Command::new(&self.cfg.shell)
                .arg("-c")
                .arg(&val)
                .current_dir(&self.work_dir)
                .envs(self.build_env())
                .output()
                .map_err(|e| ConfigError::Validation(format!("shell execution failed for var {}: {}", name, e)))?;

            if !output.status.success() {
                return Err(ConfigError::Validation(format!("shell execution failed for var {}", name)));
            }

            let result = String::from_utf8_lossy(&output.stdout).trim_end().to_string();
            self.vars.insert(name.to_string(), result);
        } else {
            self.vars.insert(name.to_string(), val);
        }
        Ok(())
    }

    fn exec_env_block(&mut self, pairs: &[crate::ast::EnvPair], body: &[FnStmt]) -> Result<(), ConfigError> {
        let mut layer = HashMap::new();
        for pair in pairs {
            let val = self.resolve_expr(&pair.value)?;
            layer.insert(pair.key.clone(), val);
        }

        let keys: Vec<&str> = pairs.iter().map(|p| p.key.as_str()).collect();
        let indent = "  ".repeat(self.env_stack.len());
        writeln!(self.writer, "{}\x1b[38;2;199;146;234menv  {}\x1b[0m", indent, keys.join(", "))
            .map_err(|e| ConfigError::Validation(format!("write error: {}", e)))?;

        self.env_stack.push(layer);

        let saved_vars = self.vars.clone();
        let result = self.exec_fn_body(body);

        self.vars = saved_vars;
        self.env_stack.pop();

        result
    }

    fn resolve_expr(&self, expr: &Expr) -> Result<String, ConfigError> {
        match expr {
            Expr::BacktickLit { parts, .. } => {
                let mut result = String::new();
                for part in parts {
                    if part.is_var {
                        let var_name = part.value.trim_start_matches('$');
                        if let Some(value) = self.vars.get(var_name) {
                            result.push_str(value);
                        } else {
                            return Err(ConfigError::Validation(format!("undefined variable: ${}", var_name)));
                        }
                    } else {
                        result.push_str(&part.value);
                    }
                }
                Ok(result)
            }
            Expr::VarRef { name, .. } => {
                if let Some(value) = self.vars.get(name) {
                    Ok(value.clone())
                } else {
                    Err(ConfigError::Validation(format!("undefined variable: ${}", name)))
                }
            }
        }
    }

    fn build_env(&self) -> Vec<(String, String)> {
        let mut env: HashMap<String, String> = HashMap::new();
        
        // Copy current environment
        for (key, value) in std::env::vars() {
            env.insert(key, value);
        }

        // Apply env stack layers
        for layer in &self.env_stack {
            for (key, value) in layer {
                env.insert(key.clone(), value.clone());
            }
        }

        env.into_iter().collect()
    }
}
